File: frontend/src/api/curing.ts
import client from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface PlantCuring {
  id_plant_curing: number
  id_plant: number
  id_session_curing: number
  date_mise_curing?: string
  date_fin_curing?: string
  poids_debut_g?: number
  poids_final_g?: number
  notes?: string
  nom_plant?: string
  nom_variete?: string
  id_culture?: number
  nom_culture?: string
  date_recolte?: string
}

export interface SessionCuring {
  id_session_curing: number
  nom?: string
  type_contenant?: string
  volume_contenant_l?: number
  boveda_rh?: number
  statut?: 'active' | 'terminee'
  date_debut?: string
  date_fin?: string
  notes?: string
  created_at?: string
  nb_plants: number
  plants: PlantCuring[]
}

export interface SessionCuringCreate {
  nom?: string
  type_contenant?: string
  volume_contenant_l?: number
  boveda_rh?: number
  id_espace?: number
  id_materiel_bocal?: number
  date_debut?: string
  notes?: string
  plants?: PlantCuringCreate[]
}

export interface SessionCuringUpdate {
  nom?: string
  type_contenant?: string
  volume_contenant_l?: number
  boveda_rh?: number
  id_espace?: number | null
  id_materiel_bocal?: number | null
  statut?: string
  date_debut?: string
  date_fin?: string
  notes?: string
}

export interface PlantCuringCreate {
  id_plant: number
  date_mise_curing?: string
  date_fin_curing?: string
  poids_debut_g?: number
  poids_final_g?: number
  notes?: string
}

export interface PlantCuringUpdate {
  date_mise_curing?: string
  date_fin_curing?: string
  poids_debut_g?: number
  poids_final_g?: number
  notes?: string
}

// ─── API ──────────────────────────────────────────────────────────────────────

export const curingAPI = {
  // Sessions
  list: (statut?: string) =>
    client.get<SessionCuring[]>('/curing/', { params: statut ? { statut } : {} }).then(r => r.data),

  get: (id: number) =>
    client.get<SessionCuring>(`/curing/${id}`).then(r => r.data),

  create: (data: SessionCuringCreate) =>
    client.post<SessionCuring>('/curing/', data).then(r => r.data),

  update: (id: number, data: SessionCuringUpdate) =>
    client.put<SessionCuring>(`/curing/${id}`, data).then(r => r.data),

  delete: (id: number) =>
    client.delete(`/curing/${id}`).then(r => r.data),

  // Plants dans une session
  addPlant: (sessionId: number, data: PlantCuringCreate) =>
    client.post<PlantCuring>(`/curing/${sessionId}/plants`, data).then(r => r.data),

  updatePlant: (sessionId: number, plantCuringId: number, data: PlantCuringUpdate) =>
    client.put<PlantCuring>(`/curing/${sessionId}/plants/${plantCuringId}`, data).then(r => r.data),

  removePlant: (sessionId: number, plantCuringId: number) =>
    client.delete(`/curing/${sessionId}/plants/${plantCuringId}`).then(r => r.data),
}

File: backend/app/schemas/curing.py
"""Schemas Pydantic pour SessionCuring et PlantCuring"""
from pydantic import BaseModel, ConfigDict
from typing import Optional
from datetime import date, datetime


# ═══════════════════════════════════════════════════════════════════════════════
# PLANT CURING
# ═══════════════════════════════════════════════════════════════════════════════

class PlantCuringCreate(BaseModel):
    id_plant: int
    date_mise_curing: Optional[date] = None
    date_fin_curing: Optional[date] = None
    poids_debut_g: Optional[float] = None
    poids_final_g: Optional[float] = None
    notes: Optional[str] = None


class PlantCuringUpdate(BaseModel):
    date_mise_curing: Optional[date] = None
    date_fin_curing: Optional[date] = None
    poids_debut_g: Optional[float] = None
    poids_final_g: Optional[float] = None
    notes: Optional[str] = None


class PlantCuringRead(BaseModel):
    id_plant_curing: int
    id_plant: int
    id_session_curing: int
    date_mise_curing: Optional[date] = None
    date_fin_curing: Optional[date] = None
    poids_debut_g: Optional[float] = None
    poids_final_g: Optional[float] = None
    notes: Optional[str] = None
    # Enrichis
    nom_plant: Optional[str] = None
    nom_variete: Optional[str] = None
    id_culture: Optional[int] = None
    nom_culture: Optional[str] = None
    date_recolte: Optional[date] = None

    model_config = ConfigDict(from_attributes=True)


# ═══════════════════════════════════════════════════════════════════════════════
# SESSION CURING
# ═══════════════════════════════════════════════════════════════════════════════

class SessionCuringCreate(BaseModel):
    nom: Optional[str] = None
    type_contenant: Optional[str] = None          # Bocal | Grove Bag | Sac sous vide | Autre
    volume_contenant_l: Optional[float] = None
    boveda_rh: Optional[int] = None               # 58 | 62 ...
    id_espace: Optional[int] = None               # espace de culture si curing dans un espace
    id_materiel_bocal: Optional[int] = None       # bocal sélectionné depuis l'inventaire
    date_debut: Optional[date] = None
    notes: Optional[str] = None
    # Plantes à ajouter directement à la création
    plants: list[PlantCuringCreate] = []


class SessionCuringUpdate(BaseModel):
    nom: Optional[str] = None
    type_contenant: Optional[str] = None
    volume_contenant_l: Optional[float] = None
    boveda_rh: Optional[int] = None
    id_espace: Optional[int] = None
    id_materiel_bocal: Optional[int] = None
    statut: Optional[str] = None
    date_debut: Optional[date] = None
    date_fin: Optional[date] = None
    notes: Optional[str] = None


class SessionCuringRead(BaseModel):
    id_session_curing: int
    nom: Optional[str] = None
    type_contenant: Optional[str] = None
    volume_contenant_l: Optional[float] = None
    boveda_rh: Optional[int] = None
    id_espace: Optional[int] = None
    id_materiel_bocal: Optional[int] = None
    # Enrichis
    nom_espace: Optional[str] = None
    nom_materiel_bocal: Optional[str] = None
    statut: Optional[str] = None
    date_debut: Optional[date] = None
    date_fin: Optional[date] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    nb_plants: int = 0
    plants: list[PlantCuringRead] = []

    model_config = ConfigDict(from_attributes=True)
